### ISSUE TITLE
libEffekseerNativeForJava.dylib has an incompatible architecture for Intel Macs

### ISSUE THREAD

**Issue by Frontear** [ChloePrime/AAAParticles#39]
Seems that according to the log the lib was built only for M1 Macs. 

Relevant log entries:

```
---- Minecraft Crash Report ----
// You should try our sister game, Minceraft!

Time: 2024-09-30 22:42:00
Description: Initializing game

java.lang.RuntimeException: Could not execute entrypoint stage 'client' due to errors, provided by 'aaa_particles'!
	at app//net.fabricmc.loader.impl.FabricLoaderImpl.lambda$invokeEntrypoints$2(FabricLoaderImpl.java:388)
	at app//net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
	at app//net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:386)
	at app//net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:53)
	at MC//net.minecraft.client.MinecraftClient.<init>(MinecraftClient.java:458)
	at net.minecraft.client.main.Main.main(Main.java:211)
	at app//net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470)
	at app//net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at app//net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
Caused by: java.lang.ExceptionInInitializerError
	at mod.chloeprime.aaaparticles.client.AAAParticlesClient.installNativeLibrary(AAAParticlesClient.java:55)
	at mod.chloeprime.aaaparticles.client.AAAParticlesClient.init(AAAParticlesClient.java:25)
	at mod.chloeprime.aaaparticles.fabric.AAAParticlesFabricClient.onInitializeClient(AAAParticlesFabricClient.java:15)
	at app//net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:384)
	... 6 more
Caused by: java.lang.UnsatisfiedLinkError: /Documents/curseforge/minecraft/Instances/Prominence II RPG_ Hasturian Era/libEffekseerNativeForJava.dylib: dlopen(/Documents/curseforge/minecraft/Instances/Prominence II RPG_ Hasturian Era/libEffekseerNativeForJava.dylib, 0x0001): tried: '/Documents/curseforge/minecraft/Instances/Prominence II RPG_ Hasturian Era/libEffekseerNativeForJava.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Documents/curseforge/minecraft/Instances/Prominence II RPG_ Hasturian Era/libEffekseerNativeForJava.dylib' (no such file), '/Documents/curseforge/minecraft/Instances/Prominence II RPG_ Hasturian Era/libEffekseerNativeForJava.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64'))
	at java.base@17.0.8/jdk.internal.loader.NativeLibraries.load(Native Method)
	at java.base@17.0.8/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
	at java.base@17.0.8/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
	at java.base@17.0.8/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
	at java.base@17.0.8/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2394)
	at java.base@17.0.8/java.lang.Runtime.load0(Runtime.java:755)
	at java.base@17.0.8/java.lang.System.load(System.java:1953)
	at mod.chloeprime.aaaparticles.client.AAAParticlesClient.installNativeLibrary(AAAParticlesClient.java:53)
	... 9 more
```

**Comment by Polaris869** [comment]
I'm honestly getting a relatively similar issue, revolving around the library for AAAparticles.


[crash-2024-10-01_18.50.44-client.txt](https://github.com/user-attachments/files/17217437/crash-2024-10-01_18.50.44-client.txt)

**Comment by Gnomecromancer** [comment]
I have the same issue

**Comment by Gnomecromancer** [comment]
Exactly the same issue

**Comment by Gnomecromancer** [comment]
Is there any plan to try to fix this issue?

**Comment by Gnomecromancer** [comment]
I would be happy to provide any help I can as a mac gamer

**Comment by Frontear** [comment]
MacOS can allegedly compile universal binaries, but I do not know if GitHub provides such through actions. Creating two dynlibs using the gh action linked by @Gnomecromancer may be a quicker fix.

**Comment by nvb-uy** [comment]
Hey @ChloePrime I recently got a macbook M3 Pro so I can build it if you need.

From my own testing, using latest version of AAAParticles it works on it too, but if needed for backwards compatibility I can compile a library.

**Comment by Leopapeo** [comment]
> Hey @ChloePrime I recently got a macbook M3 Pro so I can build it if you need.
> 
> From my own testing, using latest version of AAAParticles it works on it too, but if needed for backwards compatibility I can compile a library.

The issue is that the library libEffekseerNativeForJava.dylib that is precompiled and included as an asset in this project is a dylib that is built only for the arm64 architecture (as you have on your M3).

The ideal solution would be to replace the existing library with a fat binary, including both arm64 und x86 architectures. 

My son uses my old Intel based MacBook for playing, that's how stumbled upon tis issue. I can't figure out, what project the library originates from. I compiled the Efekseer project on my old Intel Mac, but it does not generate that library.

If @ChloePrime could point me to the source for that library, it should be trivial for me to create a fat dylib.

I also would like to offer to volunteer for testing, since I own an Intel based Mac.

**Comment by nvb-uy** [comment]
> > Hey @ChloePrime I recently got a macbook M3 Pro so I can build it if you need.
> > From my own testing, using latest version of AAAParticles it works on it too, but if needed for backwards compatibility I can compile a library.
> 
> The issue is that the library libEffekseerNativeForJava.dylib that is precompiled and included as an asset in this project is a dylib that is built only for the arm64 architecture (as you have on your M3).
> 
> The ideal solution would be to replace the existing library with a fat binary, including both arm64 und x86 architectures.
> 
> My son uses my old Intel based MacBook for playing, that's how stumbled upon tis issue. I can't figure out, what project the library originates from. I compiled the Efekseer project on my old Intel Mac, but it does not generate that library.
> 
> If @ChloePrime could point me to the source for that library, it should be trivial for me to create a fat dylib.
> 
> I also would like to offer to volunteer for testing, since I own an Intel based Mac.

The thing is that according to Frontear's investigation, ARM macbooks can compile for both ARM and x86, but GitHub actions (the way it's being compiled currently) does not have a way to do that.

**Comment by Gnomecromancer** [comment]
But someone on an arm64 computer could do the compiling, correct?

(fyi I'm very very new to all of this stuff)

**Comment by Frontear** [comment]
Yes the primary issue is that the current lib is ARM/M-series CPU compatible only. It can be compiled from the sources at https://github.com/Effekseer/EffekseerForMultiLanguages.

An ARM system should be capable of producing universal dynlibs, as per Apple documentation. I'm not sure on the specifics of how to make it do such a thing though, as in what settings it would require set.

**Comment by 4193883-eng** [comment]
i created a pull request to fix this. #44

**Comment by ChloePrime** [comment]
Temporarily fixed in 1.4.7 (make no-op instead of crashing)

**Comment by 4193883-eng** [comment]
I have updated the pull request to the universal binary

**Comment by 4193883-eng** [comment]
merged in support with #44

**Comment by ChloePrime** [comment]
Fixed in 1.4.8
